clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp

```diff
@@ -490,6 +491,13 @@ bool eligibleForExtraction(const SelectionTree::Node *N) {
         BO->getRHS() == OuterImplicit.ASTNode.get<Expr>())
       return false;
   }
+  if (const auto *Decl = Parent->ASTNode.get<VarDecl>()) {
+    if (!Decl->isInitCapture() &&
```

Comment:
it's not obvious to me why we check for anything but initializer being the outerimplict node itself. can you add comments and test cases?

Reply:
Presumably, with the BinOp check I intended to prevent a false negative for this case:
`int x = 1 + 2 + [[3 + 4 + 5]];`
Apparently, the parser considers the last "+" to be the top-level expression, so we would erroneously fail to offer extraction for the above expression.
However, the same problem already existed for the assignment case, so it's probably wrong anyway to add such a workaround for initialization only. On the other hand, I don't want to introduce a regression either. Do you have a good idea as to how a proper check could look like?

Reply:
I think I found a not-too-horrible solution.

Reply:
thanks for the example!

I guess one conceptual thing to note here is, clangd doesn't run it's refactorings on the exact selection user provided. instead it figures out an AST node that's "relevant" to this selection, and everything else is performed on that AST node, mostly independent of the selection.

So this approach is surely not always what the user means/wants, but let's us reason about the refactoring going on. For this case in particular, even if you said the refactoring is available for selection of `x = 1 + [[2 + 3]]` through these heuristics, clangd will still end up extracting the full RHS, because that's the AST node we associated with that selection.

Hence I'd actually lean towards not special casing it here, just to do the thing we're trying to prevent later on. Any particular reason why we should say extraction is available for such cases (and then still perform the extraction we're trying to avoid)?

Reply:
> For this case in particular, even if you said the refactoring is available for selection of `x = 1 + [[2 + 3]]` through 
> these heuristics, clangd will still end up extracting the full RHS, because that's the AST node we associated with that 
> selection.

I'm not sure I follow. In the example given above, invoking the tweak on the selection results in:
```
auto placeholder = 2 + 3;
int x = 1 + placeholder;
```
Which is exactly what I (the user) wanted. Am I misunderstanding something?

Reply:
Note also that there are already (pre-existing) tests for such selections, e.g.:
```
      {R"cpp(void f() {
                   int x = 1 + 2 + [[3 + 4 + 5]];
                 })cpp",
       R"cpp(void f() {
                   auto placeholder = 3 + 4 + 5; int x = 1 + 2 + placeholder;
                 })cpp"},
```

Reply:
> "end":{"character":18,"
But that's [[2 + ]]3, no? I get an end character position of 19 with both Qt Creator and VSCode for the [[2 + 3]] selection.

oh wow you're right, looks like the version of YCM I have is having off-by-one for that case.

---

sorry for the noise, i didn't know about `getBinaryOperatorRange` which does this kind of pruning. it all makes sense now.